arithmetization/src/main/java/net/consensys/linea/zktracer/container/stacked/StackedList.java

```diff
@@ -0,0 +1,134 @@
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package net.consensys.linea.zktracer.container.stacked;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+
+import com.google.common.base.Preconditions;
+import lombok.Getter;
+import lombok.experimental.Accessors;
+import net.consensys.linea.zktracer.container.ModuleOperation;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * Implements a system of nested lists behaving as a single one, where the current context
+ * modification can transparently be dropped.
+ *
+ * @param <E> the type of elements stored in the list
+ */
+@Accessors(fluent = true)
+public class StackedList<E extends ModuleOperation> {
+  private final List<E> operationSinceBeginningOfTheConflation;
+  @Getter private final List<E> thisTransactionOperation;
+  private final CountOnlyOperation lineCounter = new CountOnlyOperation();
+  private boolean conflationFinished = false;
+
+  public StackedList() {
+    operationSinceBeginningOfTheConflation = new ArrayList<>();
+    thisTransactionOperation = new ArrayList<>();
+  }
+
+  /** Prefer this constructor as we preallocate more needed memory */
+  public StackedList(
+      final int expectedConflationNumberOperations, final int expectedTransactionNumberOperations) {
+    operationSinceBeginningOfTheConflation = new ArrayList<>(expectedConflationNumberOperations);
+    thisTransactionOperation = new ArrayList<>(expectedTransactionNumberOperations);
+  }
+
+  /**
+   * when we enter a transaction, the previous transaction is definitely added to the block and
+   * can't be pop
+   */
+  public void enter() {
+    operationSinceBeginningOfTheConflation.addAll(thisTransactionOperation);
+    thisTransactionOperation.clear();
+    lineCounter.enter();
+  }
+
+  public void pop() {
+    thisTransactionOperation.clear();
+    lineCounter.pop();
+  }
+
+  public E getFirst() {
+    return operationSinceBeginningOfTheConflation.isEmpty()
+        ? thisTransactionOperation.getFirst()
+        : operationSinceBeginningOfTheConflation.getFirst();
+  }
+
+  public E getLast() {
+    return thisTransactionOperation.isEmpty()
+        ? operationSinceBeginningOfTheConflation.getLast()
+        : thisTransactionOperation.getLast();
+  }
+
+  public int size() {
+    return thisTransactionOperation.size() + operationSinceBeginningOfTheConflation.size();
+  }
+
+  public int lineCount() {
+    return lineCounter.lineCount();
+  }
+
+  public E get(int index) {
+    if (index < operationSinceBeginningOfTheConflation.size()) {
```

Comment:
This implementation looks very specific for me. Are we sure that there're no edge cases here ?

Reply:
we just want to retrieve the object at index i in a concatenation of two lists. I don't see any edge case, but I might be wrong